internal/oci/file.go

```diff
@@ -0,0 +1,264 @@
+	}
+
+	return store, nil
+}
+
+// FetchOptions configures a call to Fetch
+type FetchOptions struct {
+	IfNoMatch digest.Digest
+}
+
+// FetchResponse contains any fetched files for the given tracked reference
+// If Matched == true, then the supplied IfNoMatch digest matched and Files should be nil
+type FetchResponse struct {
+	Digest  digest.Digest
+	Files   []fs.File
+	Matched bool
+}
+
+// IfNoMatch configures the call to Fetch to return early if the supplied
+// digest matches the target manifest pointed at by the underlying reference
+// This is a cache optimization to skip re-fetching resources if the contents
+// has already been seen by the caller
+func IfNoMatch(digest digest.Digest) containers.Option[FetchOptions] {
+	return func(fo *FetchOptions) {
+		fo.IfNoMatch = digest
+	}
+}
+
+// Fetch retrieves the associated files for the tracked repository and reference
+// It can optionally be configured to skip fetching given the caller has a digest
+// that matches the current reference target
+func (s *Store) Fetch(ctx context.Context, opts ...containers.Option[FetchOptions]) (*FetchResponse, error) {
+	var options FetchOptions
+	containers.ApplyAll(&options, opts...)
+
+	desc, err := oras.Copy(ctx,
+		s.store,
+		s.reference.Reference,
+		s.local,
+		s.reference.Reference,
+		oras.DefaultCopyOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := content.FetchAll(ctx, s.local, desc)
+	if err != nil {
+		return nil, err
+	}
+
+	var manifest v1.Manifest
+	if err = json.Unmarshal(bytes, &manifest); err != nil {
+		return nil, err
+	}
+
+	var d digest.Digest
+	{
+		// shadow manifest so that we can safely
+		// strip annotations before calculating
+		// the digest
+		manifest := manifest
+		manifest.Annotations = map[string]string{}
+		bytes, err := json.Marshal(&manifest)
+		if err != nil {
+			return nil, err
+		}
+
+		d = digest.FromBytes(bytes)
+		if d == options.IfNoMatch {
+			return &FetchResponse{Matched: true, Digest: d}, nil
+		}
+	}
+
+	files, err := s.fetchFiles(ctx, manifest)
+	if err != nil {
+		return nil, err
+	}
+
+	return &FetchResponse{Files: files, Digest: d}, nil
+}
+
+// fetchFiles retrieves the associated flipt feature content files from the content fetcher.
+// It traverses the provided manifests and returns a slice of file instances with appropriate
+// content type extensions.
+func (s *Store) fetchFiles(ctx context.Context, manifest v1.Manifest) ([]fs.File, error) {
+	var files []fs.File
+
+	created, err := time.Parse(time.RFC3339, manifest.Annotations[v1.AnnotationCreated])
+	if err != nil {
+		return nil, err
+	}
+
+	for _, layer := range manifest.Layers {
+		mediaType, encoding, err := getMediaTypeAndEncoding(layer)
+		if err != nil {
+			return nil, fmt.Errorf("layer %q: %w", layer.Digest, err)
+		}
+
+		if mediaType != MediaTypeFliptNamespace {
+			return nil, fmt.Errorf("layer %q: type %q: %w", layer.Digest, mediaType, ErrUnexpectedMediaType)
+		}
```

Comment:
Hmm when would this line happen? This is if they were to point towards an image that isnt flipt related I guess?

Reply:
This would be if the OCI image was malformed. Someone had written a layer in that we dont currently expect.
We're only going to support this kind of layer for now in an OCI artifact.